ray-operator/controllers/ray/common/service.go

```diff
@@ -118,6 +118,12 @@ func BuildServiceForHeadPod(ctx context.Context, cluster rayv1.RayCluster, label
 			Type:     defaultType,
 		},
 	}
+	if defaultType == "" || defaultType == corev1.ServiceTypeClusterIP {
+		// Use headless service for the Head Pod by default,
+		// since there is no sense for Head Pods to share a same virtual IP.
+		headService.Spec.ClusterIP = corev1.ClusterIPNone
+		headService.Spec.PublishNotReadyAddresses = true // We don't need to hide the Head address if its health checks failed.
```

Comment:
Hiding them until they are ready will cause serious delays when starting up.

Reply:
> Is there any numerical difference between using `PublishNotReadyAddresses` and not using it? I also expect that there is an additional delay if we don't set `PublishNotReadyAddresses` to true. The delay mainly occurs when GCS is ready but the Ray dashboard and Raylet are not. I am curious about the extent of the difference between them.

I have tested the case of creating a RayCluster with one worker node in a fresh kind cluster:
1. with `PublishNotReadyAddresses`, the RayCluster took ~60 seconds to be fully ready, including the worker.
2. without `PublishNotReadyAddresses`, the RayCluster took ~100 seconds to be fully ready, including the worker.

~100 seconds could already make some e2e tests to fail. Both the above durations included the download time of the Ray image which took about 35 seconds.

> Why do we only set PublishNotReadyAddresses when the service type is ClusterIP?

To be more accurate, we only set `PublishNotReadyAddresses` when the service is headless. We need it to avoid immediate NXDOMAIN DNS lookup failure when the Head is not ready. The failure is the cause of the above delay.

For other service types, we will not get a DNS lookup failure since they all have a ClusterIP.

Reply:
> We need it to avoid immediate NXDOMAIN DNS lookup failure when the Head is not ready. The failure is the cause of the above delay.

Is there any evidence to support this point? I don't think this is the root cause of the delay. 

To my understanding, the delay mainly occurs when the GCS is ready, but the Ray dashboard and Raylet are not. This is because a worker Pod can register itself with the Ray head as soon as the GCS is ready. However, the head Pod is considered ready only when the GCS, Raylet, dashboard agent, and Ray dashboard are all operational.

Could you open an issue to track the progress of two items? (1) Determining the source of the delay, and (2) deciding whether we should also set `PublishNotReadyAddresses` if the service is not headless, as discussed in https://github.com/ray-project/kuberay/pull/2117/files#r1603897666.